turbopack/crates/turbo-tasks-macros-shared/src/self_filter.rs

```diff
@@ -0,0 +1,63 @@
+use proc_macro2::TokenTree;
+use quote::ToTokens;
+use syn::{
+    visit::{visit_expr, visit_macro, Visit},
+    Block, Expr, Macro,
+};
+
+pub fn is_self_used(block: &Block) -> bool {
+    let mut finder = SelfFinder { found: false };
+    finder.visit_block(block);
+    finder.found
+}
+
+struct SelfFinder {
+    found: bool,
+}
+
+impl Visit<'_> for SelfFinder {
+    fn visit_expr(&mut self, expr: &Expr) {
+        if self.found {
+            return;
+        }
```

Comment:
This probably works well because most things are expressions, but it might also be worth overriding `syn::Item` and `syn::Block` to prevent as much extra traversal as possible.